README.md
[![Open in Visual Studio Code](https://classroom.github.com/assets/open-in-vscode-f059dc9a6f8d3a56e377f745f24479a46679e63a5d9fe6f495e02850cd0d8118.svg)](https://classroom.github.com/online_ide?assignment_repo_id=6362114&assignment_repo_type=AssignmentRepo)
# a99 Finally

COMP426 final project template repository. All your code and documentation goes here. Change this text to be a brief description of your final project. Put the name in the header above. You will change everything below to be the main technical documentation, as outlined below.

## First steps

Other steps that you will need to take after you have cloned this repository:

1. Choose and update the LICENSE file accordingly. 
2. Edit this README.md file and use it as the main location of your technical documentation with links out to information contained under `/docs/`.
3. Create a `/docs/` directory for more elaborate documentation of your API, planning notes, etc.
4. Make sure that all of your team members have access to the repository as administrators.
5. Create a project under the **Projects** tab. Use this to manage your planning. Create a To-do list, etc. Explore the tools available and user them to manage your project.
7. Assign team roles and include a listing of those roles in this README.md file or in another file under `/docs/`.
8. Then put your entire development workflow in this repository.
9. Use **Pull requests** to propose changes and incorporate them into your code from various team members. 
10. Use **Issues** to identify and track bugs and also to communicate about various aspects of the project.

## Team mangement

Plan to meet with your team ASAP.
Talk through identifying roles within your team.

Try to figure out what each of you are good at/enjoy doing and try to work out roles that incorporate that.

Some basic roles you will want to consider:

1. A review manager - someone to review pull requests and merge or reject them and manage the related discussions
2. A plan manager - someone to keep an eye on the overall plan and keep the project tab/to-do list up to date
3. A documentation manager - someone to keep the documentation in order and identify what is missing and needs to be documented
4. Roles for team members to take charge or different parts of the project. Possible roles:
    1. Front end lead
    2. Back end lead
    3. Databse lead
    4. Design lead
    5. Etc.

You will notice that there are more roles than people in your group.
That is because you will all be doing a hybrid job of managing a thing while working on other things.

## Check in with instructional staff

Once all the groups are together, we'll assign each group to an LA to check in with as you progress through the project.

We will post that so that your team can schedule a time for an initial check-in with your assigned LA. It would be a good idea also to add your LA to your repository.

## Assignment instructions

And that is about all you need to get started.

All the rest of the assignment instructions are available at: https://comp426.johndmart.in/assignments/99

Good skill, and have fun with this!

# Begin Student README.md

**Dependencies**:
- better-sqlite3
- browser-sync
- concurrently
- cors
- express
- md5

**How to Run**: npm run start

docs/changes.md
# Game Changes

- can click devil as many times as you want, needs to be delay
- after final score, everything should stop
  - right now can still increase score, 0 seconds left, devil still moving
- start/end button

# Front End Changes


# Back End Changes
- form validation
- hide/show forms based on login/signup
- remove show all users form


src/app.js
const square = document.querySelectorAll('.square')
const blue_devil = document.querySelectorAll('.blue_devil')
const timeLeft = document.querySelector('#seconds-left')
let score = document.querySelector('#score')

let result = 0
let currentTime = timeLeft.textContent

function randomSquare() {
    square.forEach(className => {
        className.classList.remove('blue_devil')
    })
    
    let randomPosition = square[Math.floor(Math.random() * 9)]
    randomPosition.classList.add('blue_devil')

    //assin the id of the randomPosition to hitPosition for us to use later
    hitPosition = randomPosition.id
}

square.forEach(id => {
    id.addEventListener('mouseup', () => {
        if(id.id === hitPosition) {
            result++
            score.textContent = result
        }
    })
})

function moveDevil() {
    let timerId = null
    timerId = setInterval(randomSquare, 1000)
}
moveDevil()
function countDown() {
    currentTime--
    timeLeft.textContent = currentTime

    if (currentTime == 0) {
        clearInterval(timerId)
        alert('GAME OVER! Your final score is ' + result)

    }
}

let timerId = setInterval(countDown, 1000)


src/server.js
// Define app using express
var express = require("express");
var app = express();

// Require database SCRIPT file
var db = require("./database.js");

// Require md5 MODULE
var md5 = require("md5");

// Require CORS module
var cors = require("cors");

// Make Express use its own built-in body parser
app.use(express.urlencoded({ extended: true }));
app.use(express.json());

// Make Express use CORS
app.use(cors());

// Set server port
var HTTP_PORT = 5000;

// Start server
app.listen(HTTP_PORT, () => {
    console.log("Server running on port %PORT%".replace("%PORT%",HTTP_PORT))
});
// READ (HTTP method GET) at root endpoint /app/
app.get("/app/", (req, res, next) => {
    res.json({"message":"Your API works! (200)"});
	res.status(200);
});

// Define other CRUD API endpoints using express.js and better-sqlite3
// CREATE a new user (HTTP method POST) at endpoint /app/new/
app.post("/app/new/user", (req, res, next) => {
	var data = {
		user: req.body.user,
		email: req.body.email,
		pass: req.body.pass ? md5(req.body.pass) : null,
		name: req.body.name, 
		year: req.body.year
	}
	const stmt = db.prepare('INSERT INTO userinfo (user, pass, email, name, year) VALUES (?, ?, ?, ?, ?)');
	const info = stmt.run(data.user, data.pass, data.email, data.name, data.year);
	res.status(201).json({"message":info.changes+" record created: ID "+info.lastInsertRowid+" (201)"});
})

// READ a list of all users (HTTP method GET) at endpoint /app/users/
app.get("/app/users", (req, res) => {	
	const stmt = db.prepare("SELECT * FROM userinfo").all();
	res.status(200).json(stmt);
});


// UPDATE a single user (HTTP method PATCH) at endpoint /app/updating/user/
app.patch("/app/updating/user", (req, res) => {
	var data = {
		user: req.body.user,
		pass: req.body.pass ? md5(req.body.pass) : null,
		newuser: req.body.newuser
	}
	const stmt = db.prepare("UPDATE userinfo SET user = COALESCE(?,user) WHERE user = ? AND pass = ?")
	const info = stmt.run(data.newuser, data.user, data.pass)
	res.status(200).json({"message":`1 record updated: User ${data.user} (200)`});
});


// DELETE a single user by username and password (HTTP method DELETE) at endpoint /app/deleting/user
app.delete("/app/deleting/user", (req, res) => {
    var data = {
		user: req.body.user,
		pass: req.body.pass ? md5(req.body.pass) : null
	}
	const stmt = db.prepare("DELETE FROM userinfo WHERE user = ? AND pass = ?")
    const info = stmt.run(data.user, data.pass);
	res.status(200).json({"message":`1 record deleted: User ${data.user} (200)`});
});


// READ a single user (HTTP method GET) at endpoint /app/login/user
app.post("/app/login/user", (req, res) => {	
    var data = {
		user: req.body.user,
		pass: req.body.pass ? md5(req.body.pass) : null
	}
	const stmt = db.prepare("SELECT * FROM userinfo WHERE user = ? AND pass = ?").get(data.user, data.pass);
	res.status(200).json(stmt);
});


// READ a single user (HTTP method GET) at endpoint /app/exists/user
app.post("/app/exists/user", (req, res) => {	
    var data = {
		user: req.body.user
	}
	const stmt = db.prepare("SELECT * FROM userinfo WHERE user = ?").get(data.user);
	res.status(200).json(stmt);
});


// Default response for any other request
app.use(function(req, res){
	res.json({"message":"Your API is working!"});
    res.status(404);
});

src/database.js
// This ensures that things do not fail silently but will throw errors instead.
"use strict";
// Require better-sqlite.
const Database = require('better-sqlite3');

// Connect to a database or create one if it doesn't exist yet.
const db = new Database('src/user.db');

// Is the database initialized or do we need to initialize it?
const stmt = db.prepare(`SELECT name FROM sqlite_master WHERE type='table' and name='userinfo';`);
let row = stmt.get();
if (row === undefined) {
// Echo information about what you are doing to the console.
    console.log('Your database appears to be empty. Initializing it now.');
// Set a const that will contain your SQL commands to initialize the database.
// Include constraint that user must be unique
    const sqlInit = `
        CREATE TABLE userinfo ( id INTEGER PRIMARY KEY, user TEXT, pass TEXT, email TEXT, name TEXT, year TEXT, unique(user) );
		INSERT INTO userinfo (user, pass, email, name, year) VALUES ('admin','bdc87b9c894da5168059e00ebffb9077', 'admin@test.com', 'Sam Anthony', '2023')
    `;
// Execute SQL commands that we just wrote above.
    db.exec(sqlInit);
// Echo information about what we just did to the console.
    console.log('Your database has been initialized with a new table and three entries containing a username, email address, and password.');
} else {
// Since the database already exists, echo that to the console.
    console.log('Database exists.')
}
// Export all of the above as a module so that we can use it elsewhere.
module.exports = db

src/forms.js
window.addEventListener("load", function () {
    //  Global User Vars
    var thisUser = null;
    var loggedIn;

    // New User Creation
    function sendData( form ) {
        const sendRequest = new XMLHttpRequest();
        const signupInfo = new URLSearchParams(new FormData( form ));
        sendRequest.addEventListener("error", function(event){
            alert('Submission unsuccessful! Please try again.');
        });
        sendRequest.addEventListener("load", function(event){
            alert('Your account was created!');
        });
        sendRequest.open("POST", "http://localhost:5000/app/new/user");
        sendRequest.send( signupInfo );
    }

    function newUserValid() {
        if(document.getElementById("signup").user.value == "" | document.getElementById("signup").email.value == "" | document.getElementById("signup").pass.value == ""){
            return false;
        } else {
            return true;
        }
    }

    const newuser = document.getElementById("signup");
    newuser.addEventListener("submit", function(event){
        event.preventDefault();
        if(newUserValid()){
            sendData(this);
        } else {
            alert("Username, Email, and Password are Required!")
        }
        
    });

    // Delete User
    function deleteData( form ) {
        var deletePassword = form.pass.value;
        let deleteInfo = new URLSearchParams("");
        deleteInfo.append('user', thisUser.user);
        deleteInfo.append('pass', deletePassword);
        const sendRequest = new XMLHttpRequest();
        sendRequest.addEventListener("error", function(event){
            alert('Deletion unsuccessful! Please try again.');
        });
        sendRequest.addEventListener("load", function(event){
            // alert('Your account was deleted!');
        });

        sendRequest.open("DELETE", "http://localhost:5000/app/deleting/user");
        sendRequest.send( deleteInfo );
    }

    const olduser = document.getElementById("delete");
    olduser.addEventListener("submit", function(event){
        event.preventDefault();
        deleteData(this);
    });

    // Updating User
    function updateData( form ) {
        const sendRequest = new XMLHttpRequest();
        var updateInfo = new URLSearchParams(new FormData( form ));
        updateInfo.append('user', thisUser.user);
        sendRequest.addEventListener("error", function(event){
            alert('Changes were unsuccessful! Please try again.');
        });
        sendRequest.addEventListener("load", function(event){
            // alert('Your username was changed!');
        });
        sendRequest.open("PATCH", "http://localhost:5000/app/updating/user");
        sendRequest.send( updateInfo );
    }

    const updateUser = document.getElementById("changeName");
    updateUser.addEventListener("submit", function(event){
        event.preventDefault();
        updateData(this);
    });


    // Get List of Users
    function getUsers( form ) {
        const sendRequest = new XMLHttpRequest();
        sendRequest.addEventListener("error", function(event){
            alert('Accessing users unsuccessful! Please try again.');
        });

        sendRequest.onreadystatechange = function() {
            if (this.readyState == 4 && this.status == 200) {
                // alert(sendRequest.responseText);
                document.getElementById("allUsers").innerHTML = sendRequest.responseText;
            }
        }
        sendRequest.open("GET", "http://localhost:5000/app/users");
        sendRequest.send();        
    };

    const allUsers = document.getElementById("usersButton");
    allUsers.addEventListener("click", function(event){
        event.preventDefault();
        getUsers(this)
    });


    // Login
    function getUserdata( form ) {
        const sendRequest = new XMLHttpRequest();
        const userInfo = new URLSearchParams(new FormData( form ));
        sendRequest.addEventListener("error", function(event){
            alert('Accessing users unsuccessful! Please try again.');
        });

        sendRequest.onreadystatechange = function() {
            if (this.readyState == 4 && this.status == 200) {
                thisUser = JSON.parse(sendRequest.response);
                if(loggedIn){
                    alert("Logging Out");
                    thisUser = null;
                    loggedIn = false;
                    document.getElementById("login").value = "Login";
                    document.getElementById("greeting").innerHTML = thisUser;
                } else {
                    if(thisUser.name == null | thisUser.name == ""){
                        document.getElementById("greeting").innerHTML = "Welcome to Whack-a-Devil, " + thisUser.user;
                    } else {
                        document.getElementById("greeting").innerHTML = "Welcome to Whack-a-Devil, " + thisUser.name;
                    };
                    document.getElementById("login").value = "Logout";
                    loggedIn = true;
                    alert("Login successful!")
                }
            }
        }
        sendRequest.open("POST", "http://localhost:5000/app/login/user");
        sendRequest.send( userInfo );
    };

    const selfUser = document.getElementById("loginForm");
    selfUser.addEventListener("submit", function(event){
        event.preventDefault();
        if(loggedIn){
            alert("Logging Out");
            thisUser = null;
            loggedIn = false;
            document.getElementById("login").value = "Login";
            document.getElementById("greeting").innerHTML = thisUser;
            this.reset();
        } else {
            getUserdata(this);
        }
    });

    // Show Profile
    const showData = document.getElementById("showData");
    showData.addEventListener("click", function(event){
        event.preventDefault();
        if(loggedIn){
            document.getElementById("profileData").innerHTML = `Username: ${thisUser.user}, 
            Email: ${thisUser.email}, 
            Name: ${thisUser.name},
            Year: ${thisUser.year}`
        } else {
            alert("You must log in to see profile!")
        }
    });
});